File: movie-booking-backend/show-service/src/main/java/com/moviebooking/show/service/ShowService.java
package com.moviebooking.show.service;

import com.moviebooking.show.dtos.*;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

public interface ShowService {

    ShowResponseDto createShow(CreateShowRequestDto requestDto);

    ShowResponseDto updateShow(String showId, UpdateShowRequestDto requestDto);

    ShowResponseDto getShowById(String showId);

    List<ShowResponseDto> getShowsByMovie(String movieId);

    List<ShowResponseDto> getShowsByTheatre(String theatreId);

    List<ShowResponseDto> getShowsByScreen(String screenId);
    
    List<ShowResponseDto> getShowsByMovieAndTheatre(String movieId, String theatreId, LocalDate date);
    
    List<ShowResponseDto> getShowsByDateRange(LocalDateTime startDate, LocalDateTime endDate);

    void cancelShow(String showId);

}

File: movie-booking-frontend/src/services/showService.js
import axios from "axios";
import { API_CONFIG } from "../config/constants";
import { setupInterceptors } from "./axiosInterceptors";

const showClient = axios.create({
  baseURL: API_CONFIG.SHOW_SERVICE,
  headers: {
    "Content-Type": "application/json",
  },
});

// Setup auth token and refresh interceptors
setupInterceptors(showClient);

export const showService = {
  async getAllShows() {
    // Since there's no getAllShows endpoint, use date-range with a wide range
    const startDate = new Date("2000-01-01").toISOString().split(".")[0];
    const endDate = new Date("2099-12-31").toISOString().split(".")[0];
    try {
      const response = await showClient.get("/shows/date-range", {
        params: {
          startDate,
          endDate,
        },
      });
      return response.data || [];
    } catch (error) {
      console.warn("Could not fetch shows by date range:", error);
      return [];
    }
  },

  async getShowById(showId) {
    const response = await showClient.get(`/shows/${showId}`);
    return response.data;
  },

  async getShowsByMovie(movieId) {
    const response = await showClient.get(`/shows/movie/${movieId}`);
    return response.data;
  },

  async getShowsByTheatre(theatreId) {
    const response = await showClient.get(`/shows/theatre/${theatreId}`);
    return response.data || [];
  },

  async getShowsByMovieAndTheatre(movieId, theatreId, date = null) {
    const response = await showClient.get(
      `/shows/movie/${movieId}/theatre/${theatreId}`,
      {
        params: date ? { date } : {},
      },
    );
    return response.data || [];
  },

  async createShow(showData) {
    const response = await showClient.post("/shows", showData);
    return response.data;
  },

  async deleteShow(showId) {
    const response = await showClient.delete(`/shows/${showId}`);
    return response.data;
  },
};
